### ISSUE TITLE
Stacktrace with frames referring to anonymous classes are not grouped together

### ISSUE THREAD

**Issue by kkoch986** [getsentry/sentry-php#955]
we are seeing an issue where basically none of our issues are merging properly. As far as i can tell its due to the memory locations automatically included by PHP in the stack trace for anonymous functions:

when i go to "view similar issues" and then "diff" on sentry.io the only difference appears to be this one line in the stack trace:
![image](https://user-images.githubusercontent.com/1666071/72358698-f9252700-36ba-11ea-9a9d-dfd9b173b315.png)

ive seen the other conversations [here](https://github.com/getsentry/sentry-php/issues/481) but it seems like it was only addressed in some system areas of the code?

I would think keeping the lines in the stacktrace makes sense but maybe setting the fingerprint differently? or maybe just filtering out the memory addresses from the stack trace?

**Comment by Jean85** [comment]
As a workaround, have you tried declaring your anonymous functions as `static`, where applicable? Does this change the behavior?

BTW we extract that info from `debug_backtrace`, so it may be a bit hard for us to replace that value... 

We should improve this test to use actual data and not a fixture: https://github.com/getsentry/sentry-php/blob/9e9a5cd7a81228eef5213c26aa29d201975e4cb5/tests/StacktraceTest.php#L379-L387

@kkoch986 can you try and contribute this test for us?

**Comment by kkoch986** [comment]
could something like this work?

```php
$x = array_map(function($trace) {
            if(stripos($trace["class"] ?? null, "class@anonymous") === 0) {
                $trace["class"] = preg_replace("/0x[a-zA-Z0-9]+$/", "", $trace["class"]);
            }
            return $trace;
        }, debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS));
        var_dump($x);die();
```

takes the original trace like:

```
  [0]=>
  array(3) {
    ["function"]=>
    string(15) "messageCallback"
    ["class"]=>
    string(63) "class@anonymous^@/home/ken/etp/mq/m_q_consumer.php0x7f802ae7a419"
    ["type"]=>
    string(2) "->"
  }
  [1]=>
  array(3) {
    ["file"]=>
    string(79) "/home/ken/etp/vendor/php-amqplib/php-amqplib/PhpAmqpLib/Channel/AMQPChannel.php"
    ["line"]=>
    int(1038)
    ["function"]=>
    string(14) "call_user_func"
  }
  [2]=>
  array(3) {
    ["function"]=>
    string(13) "basic_deliver"
    ["class"]=>
    string(30) "PhpAmqpLib\Channel\AMQPChannel"
    ["type"]=>
    string(2) "->"
  }
  [3]=>
  array(3) {
    ["file"]=>
    string(83) "/home/ken/etp/vendor/php-amqplib/php-amqplib/PhpAmqpLib/Channel/AbstractChannel.php"
    ["line"]=>
    int(214)
    ["function"]=>
    string(14) "call_user_func"
  }
```
and produces this:
```
  [0]=>
  array(3) {
    ["function"]=>
    string(15) "messageCallback"
    ["class"]=>
    string(49) "class@anonymous^@/home/ken/etp/mq/m_q_consumer.php"
    ["type"]=>
    string(2) "->"
  }
  [1]=>
  array(3) {
    ["file"]=>
    string(79) "/home/ken/etp/vendor/php-amqplib/php-amqplib/PhpAmqpLib/Channel/AMQPChannel.php"
    ["line"]=>
    int(1038)
    ["function"]=>
    string(14) "call_user_func"
  }
  [2]=>
  array(3) {
    ["function"]=>
    string(13) "basic_deliver"
    ["class"]=>
    string(30) "PhpAmqpLib\Channel\AMQPChannel"
    ["type"]=>
    string(2) "->"
  }
  [3]=>
  array(3) {
    ["file"]=>
    string(83) "/home/ken/etp/vendor/php-amqplib/php-amqplib/PhpAmqpLib/Channel/AbstractChannel.php"
    ["line"]=>
    int(214)
    ["function"]=>
    string(14) "call_user_func"
  }
```

i mean to me for the sake of troubleshooting those are essentially identical, but it would allow the traces from one instance to another to be matched up

**Comment by kkoch986** [comment]
could probably do it even faster with just stripos and substr now that I'm thinking about it

**Comment by kkoch986** [comment]
in trying out one of my real stack traces in the test you referenced above seems like this is a little strange, not sure if its expected but im seeing 2 frames on the top both that seem to point to the same spot but one has no function name:

(var_dump($frames) in the UT):
```
Farray(9) {
  [0]=>
  object(Sentry\Frame)#3894 (8) {
    ["functionName":"Sentry\Frame":private]=>
    NULL
    ["file":"Sentry\Frame":private]=>
    string(25) "/path/to/app/consumer.php"
    ["line":"Sentry\Frame":private]=>
    int(498)
    ["preContext":"Sentry\Frame":private]=>
    array(0) {
    }
    ["contextLine":"Sentry\Frame":private]=>
    NULL
    ["postContext":"Sentry\Frame":private]=>
    array(0) {
    }
    ["inApp":"Sentry\Frame":private]=>
    bool(false)
    ["vars":"Sentry\Frame":private]=>
    array(0) {
    }
  }
  [1]=>
  object(Sentry\Frame)#3910 (8) {
    ["functionName":"Sentry\Frame":private]=>
    string(59) "class@anonymous/path/to/app/consumer.php0x7fc3bc369418::run"
    ["file":"Sentry\Frame":private]=>
    string(25) "/path/to/app/consumer.php"
    ["line":"Sentry\Frame":private]=>
    int(486)
    ["preContext":"Sentry\Frame":private]=>
    array(0) {
    }
    ["contextLine":"Sentry\Frame":private]=>
    NULL
    ["postContext":"Sentry\Frame":private]=>
    array(0) {
    }
    ["inApp":"Sentry\Frame":private]=>
    bool(false)
    ["vars":"Sentry\Frame":private]=>
    array(0) {
    }
  }
...
```

for this fixture:

```json
{
    "file": "path/to/file",
    "line": 12,
    "backtrace": [
        {
            "class": "class@anonymous/path/to/app/consumer.php0x7fc3bc369418",
            "function": "messageCallback",
            "type": "->"
        },
        {
            "file": "/path/to/app/vendor/php-amqplib/php-amqplib/PhpAmqpLib/Channel/AMQPChannel.php",
            "function": "call_user_func",
            "line": 1038
        },
        {
            "class": "PhpAmqpLib\\Channel\\AMQPChannel",
            "function": "basic_deliver",
            "type": "->"
        },
        {
            "file": "/path/to/app/vendor/php-amqplib/php-amqplib/PhpAmqpLib/Channel/AbstractChannel.php",
            "function": "call_user_func",
            "line": 214
        },
        {
            "class": "PhpAmqpLib\\Channel\\AbstractChannel",
            "file": "/path/to/app/vendor/php-amqplib/php-amqplib/PhpAmqpLib/Channel/AbstractChannel.php",
            "function": "dispatch",
            "line": 415,
            "type": "->"
        },
        {
            "class": "PhpAmqpLib\\Channel\\AbstractChannel",
            "file": "/path/to/app/vendor/php-amqplib/php-amqplib/PhpAmqpLib/Channel/AbstractChannel.php",
            "function": "dispatch_deferred_method",
            "line": 336,
            "type": "->"
        },
        {
            "class": "PhpAmqpLib\\Channel\\AbstractChannel",
            "file": "/path/to/app/consumer.php",
            "function": "wait",
            "line": 486,
            "type": "->"
        },
        {
            "class": "class@anonymous/path/to/app/consumer.php0x7fc3bc369418",
            "file": "/path/to/app/consumer.php",
            "function": "run",
            "line": 498,
            "type": "->"
        }
    ]
}
```

**Comment by shadowfax-chc** [comment]
I work with @kkoch986 and was also looking into this issue. It looks like in the examples @kkoch986 posted, as well as the ones I have seen, it is the `class` field in the backtrace that contains the memory address, since the it is an anonymous class.

I am not very familiar with the sentry code, but it seems like there is a spot in [add_frame] that could strip the memory address from the class:

https://github.com/getsentry/sentry-php/blob/161bd50dbae1468a165c1b3b14eda47f91f7f746/src/Stacktrace.php#L131-L137

This could even be done in a similar way as the [file prefix is stripped](https://github.com/getsentry/sentry-php/blob/161bd50dbae1468a165c1b3b14eda47f91f7f746/src/Stacktrace.php#L274-L283) with an option to determine if the memory addresses should be stripped from anonymous classes (although I can't image how it would be helpful to not group events like these).

Does this seem like the right solution?

**Comment by kkoch986** [comment]
i have a Pr almost ready with an updated test @shadowfax-chc i can try out fixing it there too

**Comment by ste93cry** [comment]
Before proceeding with any proposal or solution, are you sure that there is nothing (mainly PHP extension) that it's causing this? I honestly never saw the memory address appended to any stacktrace until now so it's a pretty new thing for me. I tried the following code and the stacktrace as you can see does not have any reference to memory addresses

```php
function foo()
{
    (function () {
        throw new Exception();
    })();
}

foo();
```

```
PHP Fatal error:  Uncaught Exception in /srv/foo.php:6
Stack trace:
#0 /test.php(7): {closure}()
#1 /test.php(10): foo()
#2 {main}
  thrown in /foo.php on line 6
```

I also tried different variations of the code by calling the lambda from outside a function, saving it to a variable first, etc and still I cannot reproduce the issue

**Comment by kkoch986** [comment]
seems to me its not from anonymous functions but anonymous classes instead. i put this together https://github.com/getsentry/sentry-php/pull/956 with a trace from our real code (with some cleaning on the files and function names).

I can try to make a smaller repeatable example.

**Comment by kkoch986** [comment]
seems like it happens when using `call_user_func` in conjunction with an anonymous class:

```php
(new class {
    public function test() {
        call_user_func([self, "run"]);
    }
    public function run() {
        var_dump(debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS));
    }
})->test();
```


```
array(2) {
  [0]=>
  array(5) {
    ["file"]=>
    string(38) "/home/ken/zentail/etp/scripts/test.php"
    ["line"]=>
    int(24)
    ["function"]=>
    string(3) "run"
    ["class"]=>
    string(68) "class@anonymous/home/ken/zentail/etp/scripts/test.php0x7f9697eb2405"
    ["type"]=>
    string(2) "->"
  }
  [1]=>
  array(5) {
    ["file"]=>
    string(38) "/home/ken/zentail/etp/scripts/test.php"
    ["line"]=>
    int(29)
    ["function"]=>
    string(4) "test"
    ["class"]=>
    string(68) "class@anonymous/home/ken/zentail/etp/scripts/test.php0x7f9697eb2405"
    ["type"]=>
    string(2) "->"
  }
}

```

**Comment by shadowfax-chc** [comment]
Just to add some more details:

If you take that same code @kkoch986 posted, but instead throw an exception, you don't see the memory addresses in the php output:

```php
(new class {
    public function test() {
        call_user_func([self, "run"]);
    }
    public function run() {
        throw new Exception();
    }
})->test();
```

```
PHP Fatal error:  Uncaught Exception in /home/tim/etp/scripts/test.php:20
Stack trace:
#0 /home/tim/etp/scripts/test.php(17): class@anonymous->run()
#1 /home/tim/etp/scripts/test.php(22): class@anonymous->test()
#2 {main}
  thrown in /home/tim/etp/scripts/test.php on line 20
```

But as @kkoch986 showed, the memory addresses are in the array returned by `debug_backtrace`.

**Comment by ste93cry** [comment]
>seems to me its not from anonymous functions but anonymous classes instead

Indeed anonymous classes are the culprit of the issue and I can now replicate the issue

>with an option to determine if the memory addresses should be stripped

Thinking about making the stripping of the memory address configurable doesn't makes much sense to me as I cannot see any usefulness of this information, so I would avoid to add a new option for this

**Comment by ste93cry** [comment]
Fixed in #956

**Comment by toemat** [comment]
Hi! I found this issue while trying to figure out why my events have not been grouped lately. I think PHP 8 changed the syntax of these anonymous class lines from `{file}0x{mem}` to `{file}:{line}${mem}`

PHP 7.x: `class@anonymous/home/test.php0x7f76c7628065`
PHP 8.x: `class@anonymous/home/test.php:3$3e0a7`

This seems to be the regex to remove the 0x style address, maybe adjust to match either 0x{hex} OR ${hex}?

https://github.com/getsentry/sentry-php/blob/64b9696d965994e0bdbd35da8e6de90bc39a269f/src/FrameBuilder.php#L79

So `/0x[a-fA-F0-9]+$/` to `/(\$|0x)[a-fA-F0-9]+$/`

(Not sure if I should open a new issue or PR? Please advise 😅)
